### PR TITLE
[v1.15] route: Specify "proto kernel" for ip routes and rules

### PR DESCRIPTION
**Pull request by jschwinger233**
* [x] #31716 @jschwinger233 

This is a customized backport to take care of upgrade/downgrade. Please see commit message for details.

### CODE REVIEW

File: pkg/proxy/routes.go [hunk 1]
```diff
@@ -196,3 +207,26 @@ func removeFromProxyRoutesIPv6() error {
...
+		if r.Protocol == unix.RTPROT_UNSPEC {
+			if err := netlink.RuleDel(&r); err != nil {
+				if !errors.Is(err, syscall.ENOENT) && !errors.Is(err, syscall.EAFNOSUPPORT) {
+					return fmt.Errorf("removing proto unspec routing rule: %w", err)
+				}
+			}
+		}
+	}
+	return nil
+}
```
Comment (julianwiedmann) [inline pkg/proxy/routes.go]: Would it help for review to split the custom cleanup logic into a separate patch (and make the first patch a clean cherry-pick from upstream) ? Or is it crucial for bisectability that all these changes go in as one patch?

Comment (jschwinger233) [reply pkg/proxy/routes.go]: Seems no harm to split into two patches. 1.15-old -> 1.15-tip upgrade is the only one that could be affected in terms of git bisect, but I don't see too many risks.

**Comment by jschwinger233** [comment]
/test-backport-1.15

File: pkg/proxy/routes.go [hunk 2]
```diff
@@ -196,3 +207,26 @@ func removeFromProxyRoutesIPv6() error {
...
+
+// removeProtoUnspecRules removes all routing rules with protocol RTPROT_UNSPEC. This is a v1.15 only function.
+func removeProtoUnspecRules(family int) error {
+	rules, err := route.ListRules(family, nil)
+	if err != nil {
+		return fmt.Errorf("listing routing rules: %w", err)
+	}
+	for _, r := range rules {
+		if r.Protocol == unix.RTPROT_UNSPEC {
```
Comment (rgo3) [inline pkg/proxy/routes.go]: Should we also ensure that we only remove rules from the tables that concern us?

Comment (jschwinger233) [reply pkg/proxy/routes.go]: Sure! Now it also checks `r.Table`:
```diff
- if r.Protocol == unix.RTPROT_UNSPEC {
+ if r.Protocol == unix.RTPROT_UNSPEC && slices.Contains(ciliumRouteTables, r.Table) {
```

**Comment by jschwinger233** [comment]
/test-backport-1.15

**Review by rgo3 — APPROVED** [review]

**Review by lmb — APPROVED** [review]
